man/man1/inetsim.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "inetsim 1"
.TH inetsim 1 "2014-05-24" "perl v5.10.1" " "
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
\&\fBinetsim\fR \- INetSim is a suite for simulating common internet services
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBinetsim\fR
[\fB\-\-config\fR <\fIconfiguration file\fR>]
[\fB\-\-version\fR]
[\fB\-\-data\-dir\fR <\fIdata directory\fR>]
[\fB\-\-log\-dir\fR <\fIlog directory\fR>]
[\fB\-\-report\-dir\fR <\fIreport directory\fR>]
[\fB\-\-bind\-address\fR <\fIip address\fR>]
[\fB\-\-max\-childs\fR <\fImaximum child processes\fR>]
[\fB\-\-user\fR <\fIuser name\fR>]
[\fB\-\-faketime\-init\-delta\fR <\fIinitial time delta\fR>]
[\fB\-\-faketime\-auto\-delay\fR <\fIdelay time\fR>]
[\fB\-\-faketime\-auto\-incr\fR <\fIstepwidth\fR>]
[\fB\-\-session\fR <\fIsession name\fR>]
[\fB\-\-pidfile\fR <\fIpid file\fR>]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBINetSim\fR simulates common internet services like
\&\fI\s-1DNS\s0\fR, \fI\s-1HTTP\s0\fR, \fI\s-1SMTP\s0\fR or \fI\s-1POP3\s0\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-\-config\fR <\fIconfiguration file\fR>" 4
.IX Item "--config <configuration file>"
The configuration file to use, default is conf/inetsim.conf in the current directory.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
Output version information.
.IP "\fB\-\-data\-dir\fR <\fIdata directory\fR>" 4
.IX Item "--data-dir <data directory>"
The data directory to use, default is data/ in the current directory.
.IP "\fB\-\-log\-dir\fR <\fIlog directory\fR>" 4
.IX Item "--log-dir <log directory>"
The log directory to use, default is log/ in the current directory.
.IP "\fB\-\-report\-dir\fR <\fIreport directory\fR>" 4
.IX Item "--report-dir <report directory>"
The report directory to use, default is report/ in the current directory.
.IP "\fB\-\-bind\-address\fR <\fI\s-1IP\s0 address\fR>" 4
.IX Item "--bind-address <IP address>"
The \s-1IP\s0 address to bind services to.
.IP "\fB\-\-max\-childs\fR <\fImaximum child processes\fR>" 4
.IX Item "--max-childs <maximum child processes>"
The maximum number of child processes (number of parallel connections) for
each service. Default is 10.
.IP "\fB\-\-user\fR <\fIusername\fR>" 4
.IX Item "--user <username>"
User to run services. Default is 'nobody'.
.IP "\fB\-\-faketime\-init\-delta\fR <\fIinitial time delta\fR>" 4
.IX Item "--faketime-init-delta <initial time delta>"
Initial number of seconds (positive or negative) relative to current
date/time for fake time used by all services. This overrides the
option 'faketime\-init\-delta' in the configuration file.
Default is 0 (use current date/time).
.IP "\fB\-\-faketime\-auto\-delay\fR <\fIdelay time\fR>" 4
.IX Item "--faketime-auto-delay <delay time>"
Number of seconds to wait before incrementing or decrementing fake time
by amount of seconds specified with \fBfaketime-auto-incr\fR.
Setting to '0' disables this option. This overrides the option
\&'faketime\-auto\-delay' in the configuration file.
Default is 0 (disabled).
.IP "\fB\-\-faketime\-auto\-incr\fR <\fIstepwidth\fR>" 4
.IX Item "--faketime-auto-incr <stepwidth>"
Number of seconds by which fake time is incremented or decremented at
regular intervals specified by \fBfaketime_auto_delay\fR. This option only
takes effect if \fBfaketime-auto-delay\fR is enabled (not set to '0').
This overrides the option 'faketime\-auto\-incr' in the configuration file.
Default is 3600.
.IP "\fB\-\-session\fR <\fIsession name\fR>" 4
.IX Item "--session <session name>"
Session name to use, default is the \s-1PID\s0 of the parent process.
.IP "\fB\-\-pidfile\fR <\fIpid file\fR>" 4
.IX Item "--pidfile <pid file>"
The pid file to use, default is /var/run/inetsim.pid.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Start the simulation with session name 'simtest5'
.PP
.Vb 1
\&        # inetsim \-\-session simtest5
.Ve
.PP
Start the simulation with a one-day-jump to the future after 60 seconds
.PP
.Vb 1
\&        # inetsim \-\-faketime\-auto\-delay 60 \-\-faketime\-auto\-incr 86400
.Ve
.PP
dito, but with jump to the past
.PP
.Vb 1
\&        # inetsim \-\-faketime\-auto\-delay 60 \-\-faketime\-auto\-incr \-86400
.Ve
.SH "REQUIRES"
.IX Header "REQUIRES"
Perl 5, Getopt::Long, Net::Server, Net::DNS, IO::Handle, IO::Socket,
IO::Select, IPC::Shareable, Digest::SHA, File::Copy, MIME::Base64,
IPTables::IPv4::IPQueue (optional)
.SH "AUTHOR"
.IX Header "AUTHOR"
Matthias\ Eckert <me@inetsim.org>, 
Thomas\ Hungenberg <th@inetsim.org>


CHANGES
1.2.5 (2014-05-24)
------------------
- Fixed incompatibility issue with newer versions of IO::Socket::SSL
- Fixed incompatibility issue with some NTP clients
- Set HTTP(S) POST data directory correctly if '--data-dir' command
  line option is used


1.2.4 (2013-08-15)
------------------
- Fixed broken report module


1.2.3 (2012-10-01)
------------------
- Changed Debian package dependency from Digest::SHA1 to Digest::SHA
- Changed process name format


1.2.2 (2010-11-24)
------------------
- DNS:
  - added check for broken version 0.65 of Net::DNS
- HTTP:
  - added support for static fakefiles
  - some bugfixes with HTTPS fakemode


1.2.1 (2010-07-09)
------------------
- POP3: fixed a small typo that caused the module to be practically unusable


1.2 (2010-04-25)
----------------
- new service module: IRC
  - basic command set
- SMTP:
  - added SSL support (smtps)
  - added support for service extension STARTTLS
  - added support for enforced authentication
- POP3:
  - added SSL support (pop3s)
  - added support for capabilities STLS, TOP, IMPLEMENTATION, LOGIN-DELAY, EXPIRE,
    RESP-CODES and AUTH-RESP-CODE
- HTTP:
  - added SSL support (https)
- FTP:
  - added SSL support (ftps)
- TFTP:
  - added support for TFTP options 'blksize', 'timeout' and 'tsize'
  - added virtual file system
- Redirect:
  - added support for ICMP
- some minor bugfixes


1.1.1 (2009-09-09)
------------------
- FTP: always re-read content of static ftp root directory on startup
- SMTP: set 'Return-Path' to '<>' for MAILER-DAEMON messages,
        replaced header 'Delivered-To' with 'Envelope-To'
- POP3: fixed logging of non-printable characters
- some minor bugfixes


1.1 (2008-10-12)
----------------
- new service module: FTP
- new service module: Syslog
- added support for HTTP methods POST and OPTIONS
- added support for port numbers in HTTP 'Host' header and absoluteURI
- changed SMTP HELO/EHLO response
- added support for SMTP service extensions ENHANCEDSTATUSCODES,
  SEND, SAML, SOML, TURN, ETRN, ATRN, MTRK, BINARYMIME,
  CHUNKING, DELIVERBY, SUBMITTER, CHECKPOINT, NO-SOLICITING, FUTURERELEASE
- added configuration options 'dummy_banner' and 'dummy_banner_wait'
- added configuration options 'redirect_ignore_bootp' and 'redirect_ignore_netbios'
- added configuration options 'create_reports' and 'report_language'
- changed format of service logfile to include service child process id
- changed default value for 'service_timeout' to 120 seconds
- install script 'set_permissions.sh' renamed to 'setup.sh'


1.0 (2008-07-07)
----------------
- changed default setting for configuration option 'redirect_enabled' to 'no'


1.0rc4 (2008-06-29)
-------------------
- Perl library Net::Server is no longer included with the INetSim distribution
- added 'dummy' service module (just logs all received data)
- added 'redirect' feature:
  - uses Linux kernel IP_QUEUE, so only available on Linux platforms
  - allows for IP-based connection redirection (tcp/udp)
  - supports static rules for redirection based on target IP address, port and/or protocol
  - can act as a NAT router for redirection of packets to other hosts
  - optionally varies the TTL value of IP packets sent to the clients from different
    "virtual" connection targets to make traffic look more authentic
- DNS.pm: added handling of AAAA queries (returns empty answer with status 'NOERROR')
- SMTP.pm: add 'Received:' headers to messages
- HTTP.pm: improved Request-URI parsing, now supports absolute URIs
- NTP.pm: added configuration option for less strict checks
- added timeout handling for all service modules
- removed 'bind_address' configuration options for individual services
- some minor bugfixes


1.0rc3 (2007-12-12)
-------------------
- added 'finger' service module
- added configuration option for SMTP and POP3 to select if only
  reversible or all supported authentication mechanisms are offered
- added support for SMTP extensions DSN, ETRN, EXPN, HELP and VERP


1.0rc2 (2007-10-21)
-------------------
- added TFTP service module
- added POP3 configuration options pop3_mbox_maxmails, pop3_mbox_reread
  and pop3_mbox_rebuild
- added sample SPAM mbox file for POP3


1.0rc1 (2007-10-17)
-------------------
- first public release


lib/INetSim.pm
# -*- perl -*-
#
# INetSim - An internet simulation framework
#
# (c)2007-2014 Matthias Eckert, Thomas Hungenberg
#
# Version 1.2.5 (2014-05-24)
#
# For history/changelog see bottom of this file.
#
#############################################################

package INetSim;

use strict;
use warnings;
use POSIX;
# modules to use
use INetSim::CommandLine;
use INetSim::Config;
use INetSim::Log;
use INetSim::FakeTime;
use INetSim::Chargen::TCP;
use INetSim::Chargen::UDP;
use INetSim::Daytime::TCP;
use INetSim::Daytime::UDP;
use INetSim::Discard::TCP;
use INetSim::Discard::UDP;
use INetSim::Echo::TCP;
use INetSim::Echo::UDP;
use INetSim::Quotd::TCP;
use INetSim::Quotd::UDP;
use INetSim::Time::TCP;
use INetSim::Time::UDP;
use INetSim::HTTP;
use INetSim::Ident;
use INetSim::NTP;
use INetSim::SMTP;
use INetSim::POP3;
use INetSim::DNS;
use INetSim::TFTP;
use INetSim::Report;
use INetSim::Finger;
use INetSim::Dummy::TCP;
use INetSim::Dummy::UDP;
use INetSim::FTP;
use INetSim::Syslog;
use INetSim::IRC;

my $VERSION = "INetSim 1.2.5 (2014-05-24)";


#############################################################
# Local variables

my $PPID = $$;   # Parent PID
my @childs = (); # Child PIDs


#############################################################
# Child process handling
#
sub fork_services {
    my @services_to_start = &INetSim::Config::getServicesToStart();
    foreach (@services_to_start) {
	my $pid = fork();
	if ($pid) {
	    # we are the parent process
	    push(@childs, $pid);
	}
	elsif ($pid == 0){
	    # we are the child process
	    if(/^dns$/) {
		&INetSim::DNS::dns;
	    }
	    elsif(/^smtp$/) {
		INetSim::SMTP->run;
	    }
	    elsif(/^smtps$/) {
		INetSim::SMTP->new({ SSL => 1 })->run;
	    }
	    elsif(/^pop3$/) {
		INetSim::POP3->run;
	    }
	    elsif(/^pop3s$/) {
		INetSim::POP3->new({ SSL => 1 })->run;
	    }
	    elsif(/^http$/) {
		INetSim::HTTP->run;
	    }
	    elsif(/^https$/) {
		INetSim::HTTP->new({ SSL => 1 })->run;
	    }
	    elsif(/^ntp$/) {
		INetSim::NTP->run;
	    }
	    elsif(/^time_tcp$/) {
		INetSim::Time::TCP->run;
	    }
	    elsif(/^time_udp$/) {
		INetSim::Time::UDP->run;
	    }
	    elsif(/^daytime_tcp$/) {
		INetSim::Daytime::TCP->run;
	    }
	    elsif(/^daytime_udp$/) {
		INetSim::Daytime::UDP->run;
	    }
	    elsif(/^ident$/) {
		INetSim::Ident->run;
	    }
	    elsif(/^echo_tcp$/) {
		INetSim::Echo::TCP->run;
	    }
	    elsif(/^echo_udp$/) {
		INetSim::Echo::UDP->run;
	    }
	    elsif(/^discard_tcp$/) {
		INetSim::Discard::TCP->run;
	    }
	    elsif(/^discard_udp$/) {
		INetSim::Discard::UDP->run;
	    }
	    elsif(/^chargen_tcp$/) {
		INetSim::Chargen::TCP->run;
	    }
	    elsif(/^chargen_udp$/) {
		INetSim::Chargen::UDP->run;
	    }
	    elsif(/^quotd_tcp$/) {
		INetSim::Quotd::TCP->run;
	    }
	    elsif(/^quotd_udp$/) {
		INetSim::Quotd::UDP->run;
	    }
	    elsif(/^tftp$/) {
		INetSim::TFTP->run;
	    }
	    elsif(/^finger$/) {
		INetSim::Finger->run;
	    }
	    elsif(/^dummy_tcp$/) {
		INetSim::Dummy::TCP->run;
	    }
	    elsif(/^dummy_udp$/) {
		INetSim::Dummy::UDP->run;
	    }
	    elsif(/^ftp$/) {
		INetSim::FTP->run;
	    }
	    elsif(/^ftps$/) {
		INetSim::FTP->new({ SSL => 1 })->run;
	    }
	    elsif(/^syslog$/) {
		INetSim::Syslog->run;
	    }
	    elsif(/^irc$/) {
		INetSim::IRC->run;
	    }
	    elsif(/^ircs$/) {
		INetSim::IRC->run( SSL => 1 );
	    }
	}
	else {
	    &error_exit ("Could not fork: $!", 1);
	}
    }
    sleep 1;
}


sub handle_pid {
    my $cmd = shift;
    my $pidfile = &INetSim::CommandLine::getCommandLineOption("pidfile");

    $pidfile =~ /(.*)/; # evil untaint
    $pidfile = $1;
    if ($cmd eq "create") {
        if (-f $pidfile) {
	    print STDOUT "PIDfile '$pidfile' exists - INetSim already running?\n";
	    exit 1;
	}
	else {
	    if (! open (PIDFILE, "> $pidfile")) {
		print STDOUT "Unable to open PIDfile for writing: $!\n";
		exit 1;
	    }
	    print PIDFILE $PPID;
	    close PIDFILE;
	}
    }
    elsif ($cmd eq "remove") {
        if (-f $pidfile) {
	    unlink $pidfile;
	}
	else {
	    print STDOUT "Hmm, PIDfile '$pidfile' not found (but, who cares?)\n";
	}
    }
}


sub auto_faketime {
    if (&INetSim::Config::getConfigParameter("Faketime_AutoDelay") > 0) {
	my $pid = fork();
	if ($pid) {
	    # we are the parent process
	    push(@childs, $pid);
	}
	elsif ($pid == 0){
	    # we are the child process
	    &INetSim::FakeTime::auto_faketime();
	}
    }
}


sub redirect_packets {
    if (&INetSim::Config::getConfigParameter("Redirect_Enabled")) {
        # check for linux
        if ($^O !~ /linux/i) {
            &INetSim::Log::MainLog("failed! Error: Sorry, the Redirect module does not support this operating system!", "redirect");
            return 0;
        }
        # check for Perlipq library
        eval {
               eval "use IPTables::IPv4::IPQueue; 1" or die;
        };
        if ($@) {
            &INetSim::Log::MainLog("failed! Error: Sorry, this module requires the Perlipq library (IPTables::IPv4::IPQueue)!", "redirect");
            return 0;
        }
        # check for redirect module
        eval {
               eval "use INetSim::Redirect; 1" or die;
        };
        if ($@) {
            &INetSim::Log::MainLog("failed! Error: $@", "redirect");
            return 0;
        }
        my $pid = fork();
        if ($pid) {
	    # we are the parent process
	    push(@childs, $pid);
        }
        elsif ($pid == 0){
	    # we are the child process
	    &INetSim::Redirect::run();
        }
    }
}


sub rest_in_peace {
    my $count = @childs;
    my $i;

    for ($i = 0; $i < $count; $i++) {
        waitpid(-1,&WNOHANG);
        if (! (kill (0, $childs[$i]))) {
            splice (@childs, $i, 1);
            $count = @childs;
            $i--;
        }
    }
}


sub wait_pids {
    wait();
    foreach (@childs){
	waitpid($_, 0);
    }
}


sub kill_pids {
    foreach (@childs){
	kill("TERM", $_);
	waitpid($_, 0);
    }
}


sub error_exit {
    my $msg = shift;
    if (! defined $msg) {
	$msg = "Unknown error";
    }
    my $exitcode = shift;
    if (! defined $exitcode) {
	$exitcode = 1;
    }
    elsif (($exitcode !~ /^[\d]{1,3}$/) || (int($exitcode) < 0) || (int($exitcode > 255))) {
	print STDOUT "Illegal exit code!\n";
	$exitcode = 1;
    }

    print STDOUT "Error: $msg.\n";

    &kill_pids;
    &wait_pids;
    &handle_pid("remove");

    exit 1;
}


#############################################################
# Main
#

sub main {
    # Parse commandline options
    &INetSim::CommandLine::parse_options();

    # Check command line option 'help'
    if (&INetSim::CommandLine::getCommandLineOption("help")) {
	print STDOUT << "EOF";
$VERSION by Matthias Eckert & Thomas Hungenberg

Usage: $0 [options]

Available options:
  --help                         Print this help message.
  --version                      Show version information.
  --config=<filename>            Configuration file to use.
  --log-dir=<directory>          Directory logfiles are written to.
  --data-dir=<directory>         Directory containing service data.
  --report-dir=<directory>       Directory reports are written to.
  --bind-address=<IP address>    Default IP address to bind services to.
                                 Overrides configuration option 'default_bind_address'.
  --max-childs=<num>             Default maximum number of child processes per service.
                                 Overrides configuration option 'default_max_childs'.
  --user=<username>              Default user to run services.
                                 Overrides configuration option 'default_run_as_user'.
  --faketime-init-delta=<secs>   Initial faketime delta (seconds).
                                 Overrides configuration option 'faketime_init_delta'.
  --faketime-auto-delay=<secs>   Delay for auto incrementing faketime (seconds).
                                 Overrides configuration option 'faketime_auto_delay'.
  --faketime-auto-incr=<secs>    Delta for auto incrementing faketime (seconds).
                                 Overrides configuration option 'faketime_auto_increment'.
  --session=<id>                 Session id to use. Defaults to main process id.
  --pidfile=<filename>           Pid file to use. Defaults to '/var/run/inetsim.pid'.

EOF
;
	exit 0;
    }
    elsif (&INetSim::CommandLine::getCommandLineOption("version")) {
        print STDOUT "$VERSION by Matthias Eckert & Thomas Hungenberg\n";
        exit 0;
    }


    # Check if we are running with root privileges (EUID 0)
    if ( $> != 0 ) {
	print STDOUT "Sorry, this program must be started as root!\n";
	exit 1;
    }

    # Check if group "inetsim" exists on system
    my $gid = getgrnam("inetsim");
    if (! defined $gid) {
	print STDOUT "No such group 'inetsim' configured on this system!\n";
	print STDOUT "Please create group and start again. See documentation for more information.\n";
	exit 1;
    }

    print STDOUT "$VERSION by Matthias Eckert & Thomas Hungenberg\n";

    # create pidfile
    &handle_pid("create");

    # Parse configuration file
    &INetSim::Config::parse_config;

    # Check if there are services to start configured, else exit
    if (! scalar(&INetSim::Config::getServicesToStart())) {
	&INetSim::Log::MainLog("No services to start configured. Exiting.");
	&handle_pid("remove");
	exit 0;
    }

    # ignore some signal handlers during startup
    local $SIG{'INT'} = 'IGNORE';
    local $SIG{'HUP'} = 'IGNORE';
    local $SIG{'TERM'} = 'IGNORE';

    &INetSim::Log::MainLog("=== INetSim main process started (PID $PPID) ===");
    &INetSim::Log::MainLog("Session ID:     " . &INetSim::Config::getConfigParameter("SessionID"));
    &INetSim::Log::MainLog("Listening on:   " . &INetSim::Config::getConfigParameter("Default_BindAddress"));
    &INetSim::Log::MainLog("Real Date/Time: " . strftime "%Y-%m-%d %H:%M:%S", localtime);
    &INetSim::Log::MainLog("Fake Date/Time: " . (strftime "%Y-%m-%d %H:%M:%S", localtime(&INetSim::FakeTime::get_faketime())). " (Delta: " . &INetSim::Config::getConfigParameter("Faketime_Delta") . " seconds)");
    &INetSim::Log::MainLog(" Forking services...");
    &fork_services();
    &auto_faketime();
    &redirect_packets();

    if ($$ == $PPID) {
	$0 = 'inetsim_main';
	sleep 2;
        # reap zombies ;-)
        &rest_in_peace;
	&INetSim::Log::MainLog(" done.");
	&INetSim::Log::MainLog("Simulation running.");
	# catch up some signalhandlers for the parent process
	local $SIG{'INT'} = sub {&kill_pids;};
	local $SIG{'HUP'} = sub {&kill_pids;};
	local $SIG{'TERM'} = sub {&kill_pids;};
	&wait_pids;
	&INetSim::Log::MainLog("Simulation stopped.");

	# create report
	if (&INetSim::Config::getConfigParameter("Create_Reports")) {
	    &INetSim::Report::GenReport;
	}

	&INetSim::Log::MainLog("=== INetSim main process stopped (PID $PPID) ===");
	&INetSim::Log::MainLog(".");
    }

    # delete pidfile
    &handle_pid("remove");
    exit 0;
}


1;
#############################################################
#
# History:
#
# Version 1.2.4 (2013-08-14) th
# - changed date/time output format
#
# Version 1.2beta5 (2010-04-15) me
# - added 'Default_BindAddress' to startup screen
#
# Version 1.2beta4 (2009-12-15) th
#
# Version 1.2beta3 (2009-09-23) me  [branch]
# - added service IRC/IRCs
#
# Version 1.1.1 (2009-09-09) th
# - 1.1.1 release
#
# Version 1.2beta2 (2009-09-04) me  [branch]
# - services POP3, HTTP and FTP prepared for using SSL too
#
# Version 1.2beta1 (2009-09-03) me  [branch]
# - added service SMTPS
# - added commandline option '--version' to help output
#
# Version 1.1 (2008-10-12) me
# - 1.1 release
#
# Version 1.1pre4 (2008-09-08) me
# - added service syslog
#
# Version 1.1pre3 (2008-08-24) me
# - changed FTP module name
#
# Version 1.1pre2 (2008-08-20) me
# - added service FTP
#
# Version 1.1pre1 (2008-08-09) th
# - added configuration option 'Create_Reports'
#
# Version 1.0 (2008-07-06) th
# - 1.0 release
#
# Version 1.0rc4 (2008-06-26) th
# - 1.0 rc4 release
#
# Version 1.0rc4pre9 (2008-03-20) me
# - fixed checks in function redirect_packets()
#
# Version 1.0rc4pre8 (2008-03-19) me
# - changed handling for OS check and added an error message
# - removed check for use of 'INetSim::Redirect' (see line below)
# - added check for use of 'IPTables::IPv4::IPQueue' instead of
#   the redirect module and added an error message
# - added function rest_in_peace() to get possible zombies
#
# Version 1.0rc4pre7 (2008-03-17) me
# - added &INetSim::Config::getConfigParameter("Redirect_Enabled") in
#   function redirect_packets()
#
# Version 1.0rc4pre6 (2008-03-15) me
# - added check for use of 'INetSim::Redirect', because it's
#   system dependent
# - added check for operating system in function redirect_packets()
#
# Version 1.0rc4pre5 (2008-03-07) me
# - added service "redirect"
#
# Version 1.0rc4pre4 (2008-03-06) me
# - added service "dummy"
#
# Version 1.0rc4pre3 (2008-03-05) me
# - moved checks for uid '0' and group 'inetsim' below commandline
#   parser, because '--help' should always be possible
#
# Version 1.0rc4pre2 (2008-02-17) me
# - replaced variable 'pidfile' with
#   &INetSim::CommandLine::getCommandLineOption("pidfile");
# - unused old code in function 'handle_pid' removed
#
# Version 1.0rc4pre1 (2007-12-31) th
# - change process names
#
# Version 1.0rc3 (2007-12-12) me
# - new public release
#
# Version 1.0rc3pre1 (2007-11-07) me
# - added service finger
#
# Version 1.0rc2 (2007-10-21) th
# - new public release
#
# Version 1.0rc1 (2007-10-17) th
# - first public release
#
# Version 0.41   (2007-05-31) th
# - added/removed some comments
#
# Version 0.40   (2007-05-23) th
# - added 'sleep 1' to main process to wait for all services started
#
# Version 0.39   (2007-04-30) th
# - check if group 'inetsim' exists on system
# - use getServicesToStart() instead of accessing module
#   variables in Config.pm
#
# Version 0.38   (2007-04-29) th
# - added module CommandLine.pm
# - added check for command line option 'help'
# - added global variable 'VERSION'
#
# Version 0.37   (2007-04-25) th
# - replaced &INetSim::FakeTime::get_faketime_delta() with
#   &INetSim::Config::getConfigParameter("faketime_delta")
#
# Version 0.36   (2007-04-22) th
# - added function error_exit()
# - renamed parameter options in handle_pid() to "create" and "remove"
#
# Version 0.35   (2007-04-21) me
# - added GenReport()
#
# Version 0.34   (2007-04-20) me
# - added check for uid 0
#
# Version 0.33   (2007-04-20) me
# - added function for pidfile handling
#
# Version 0.32   (2007-04-20) me
# - added logging of $INetSim::Config::SessionID
# - re-added a dot at the end of main log
#   (this makes the log easier to read)
#
# Version 0.31   (2007-04-10) th
# - renamed fork_servers() to fork_services()
# - added auto_faketime()
#
# Version 0.30   (2007-04-09) th
# - added module FakeTime
#
# Version 0.30   (2007-04-05) th
# - changed module name 'dns.pm' to 'DNS.pm'
#
# Version 0.29   (2007-03-30) me
# - added service tftp
#
# Version 0.28   (2007-03-27) th
# - moved CONFIGFILENAME, MAINLOGFILENAME and
#   SUBLOGFILENAME to INetSim::Config
# - moved logging functions to INetSim::Log
#
# Version 0.27   (2007-03-26) th
# - changed daytime_tcp to use INetSim::GenericServer
# - changed daytime_udp to use INetSim::GenericServer
# - changed time_tcp to use INetSim::GenericServer
# - changed time_udp to use INetSim::GenericServer
# - changed quotd_tcp to use INetSim::GenericServer
# - changed quotd_udp to use INetSim::GenericServer
# - changed discard_tcp to use INetSim::GenericServer
# - changed discard_udp to use INetSim::GenericServer
# - changed smtp to use INetSim::GenericServer
# - changed pop3 to use INetSim::GenericServer
#
# Version 0.26   (2007-03-24) th
# - changed chargen_tcp to use INetSim::GenericServer
# - changed chargen_udp to use INetSim::GenericServer
# - changed http to use INetSim::GenericServer
#
# Version 0.25   (2007-03-23) th
# - changed echo_tcp to use INetSim::GenericServer
# - changed echo_udp to use INetSim::GenericServer
#
# Version 0.24   (2007-03-19) th
# - added service http
#
# Version 0.23   (2007-03-19) me
# - added service ident
# - added service echo
# - added service discard
# - added service chargen
# - added service discard
#
# Version 0.22   (2007-03-16) th
# - added configuration option @INetSim::Config::ServicesToStart
# - changed BASEDIR to "."
#
# Version 0.21   (2007-03-15) me
# - ignore signals during startup
#
# Version 0.2    (2007-03-15) th
# - added configuration module
# - rewrote fork_servers()
# - changed logfile location
#
# Version 0.1    (2007-03-12) me
#
#############################################################
